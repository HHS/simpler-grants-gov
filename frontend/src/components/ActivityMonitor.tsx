"use client";

import { useUser } from "src/services/auth/useUser";

import { useCallback, useEffect, useRef, useState } from "react";

// when a user is logged in, this will track all clicks and key presses and,
// on each action, if the user's auth token is expiring or expired, will refresh the
// token or log the user out accordingly
export function ActivityMonitor() {
  const { user, refreshIfExpiring, refreshIfExpired, logoutLocalUser } =
    useUser();

  const [listening, setListening] = useState(false);
  const handlerRef = useRef<EventListener>(null);

  const refreshTokenIfExpiringOrLogout = useCallback(() => {
    refreshIfExpired()
      .then((expired) => {
        if (expired) {
          logoutLocalUser();
          return;
        }
        return refreshIfExpiring();
      })
      .catch((e) => {
        console.error("Error refreshing token in response to user activity", e);
      });
  }, [refreshIfExpired, refreshIfExpiring, logoutLocalUser]);

  const addHandlers = useCallback(() => {
    if (!listening) {
      document.addEventListener("click", refreshTokenIfExpiringOrLogout);
      document.addEventListener("keydown", refreshTokenIfExpiringOrLogout);
      setListening(true);
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  const removeHandlers = useCallback(() => {
    if (!listening) {
      return;
    }
    document.removeEventListener("click", handlerRef.current as EventListener);
    document.removeEventListener(
      "keydown",
      handlerRef.current as EventListener,
    );
    setListening(false);
    handlerRef.current = null;
  }, [listening]);

  // when logged in status changes, remove or add handlers
  useEffect(() => {
    if (!user || !user.token) {
      removeHandlers();
      return;
    }
    addHandlers();
    return () => {
      removeHandlers();
    };
  }, [user, addHandlers, removeHandlers]);

  // whenever we are not listening for activity, and the handler function has been updated
  // update the ref, so that we have the right reference ready when we want to remove the handlers later.
  // we do not want to pick up any changes made while listening, as those changes will not corerspond to the
  // handler that is actually attached to the document.
  // note that if we try to do this in addHandlers it somehow doesn't work right, given all of the dependencies involved.
  useEffect(() => {
    if (refreshTokenIfExpiringOrLogout && !listening) {
      handlerRef.current = refreshTokenIfExpiringOrLogout;
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  return null;
}
