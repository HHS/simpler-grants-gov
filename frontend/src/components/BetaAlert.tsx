import { useTranslations } from "next-intl";
import { ReactNode } from "react";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BetaAlertProps = {
  containerClasses?: string;
  heading?: string;
  alertMessage?: string | ReactNode;
};

const BetaAlert = ({
  containerClasses,
  heading,
  alertMessage,
}: BetaAlertProps) => {
  const t = useTranslations("BetaAlert");
  const defaultAlertMessage = t.rich("alert", {
    LinkToGrants: (content) => <a href="https://www.grants.gov">{content}</a>,
  });

  return (
    <div data-testid="beta-alert" className={containerClasses}>
      <GridContainer>
        <USWDSAlert
          type="warning"
          headingLevel="h2"
          heading={heading || t("alertTitle")}
          noIcon
        >
          {alertMessage || defaultAlertMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BetaAlert;
