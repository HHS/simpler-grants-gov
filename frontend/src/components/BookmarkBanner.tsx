import { useTranslations } from "next-intl";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BookmarkBannerProps = {
  containerClasses?: string;
  message?: string;
};

const BookmarkBanner = ({ containerClasses, message }: BookmarkBannerProps) => {
  const t = useTranslations("BookmarkBanner");

  const defaultMessage = (
    <>
      {t("message")}
      <br />
      {t.rich("technicalSupportMessage", {
        mailToGrants: (content) => (
          <a href="mailto:simpler@grants.gov">{content}</a>
        ),
      })}
    </>
  );

  return (
    <div data-testid="bookmark-banner" className={containerClasses}>
      <GridContainer>
        <USWDSAlert type="info" headingLevel="h2" heading={t("title")}>
          {message || defaultMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BookmarkBanner;
