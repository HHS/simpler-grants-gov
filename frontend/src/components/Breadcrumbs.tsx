import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  GridContainer,
} from "@trussworks/react-uswds";

export type Breadcrumb = {
  title: string;
  path: string;
};

export type BreadcrumbList = Breadcrumb[];

type Props = {
  breadcrumbList: BreadcrumbList;
};

const microdata = {
  ol: {
    itemScope: true,
    itemType: "https://schema.org/BreadcrumbList",
  },
  li: {
    itemScope: true,
    itemProp: "itemListElement",
    itemType: "https://schema.org/ListItem",
  },
  a: {
    itemScope: true,
    itemProp: "item",
    itemType: "https://schema.org/WebPage",
  },
};

const Breadcrumbs = ({ breadcrumbList }: Props) => {
  const breadcrumArray = breadcrumbList.map((breadcrumbInfo, i) => {
    return (
      <Breadcrumb
        key={breadcrumbInfo.title + "-crumb"}
        current={i + 1 === breadcrumbList.length}
        {...microdata.li}
      >
        {i + 1 !== breadcrumbList.length ? (
          <BreadcrumbLink
            href={breadcrumbInfo.path}
            itemID={breadcrumbInfo.path}
            {...microdata.a}
          >
            {}
            <span itemProp="name">{breadcrumbInfo.title}</span>
          </BreadcrumbLink>
        ) : (
          <span itemProp="name">{breadcrumbInfo.title}</span>
        )}
        <meta itemProp="position" content={(i + 1).toString()} />
      </Breadcrumb>
    );
  });

  return (
    <GridContainer data-testid="breadcrumb">
      <BreadcrumbBar listProps={{ ...microdata.ol }}>
        {breadcrumArray}
      </BreadcrumbBar>
    </GridContainer>
  );
};

export default Breadcrumbs;
