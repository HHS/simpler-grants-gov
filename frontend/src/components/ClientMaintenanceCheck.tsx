"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";

import { useEffect } from "react";

/**
 * Checks for the "major" features being offline, otherwise we're not in maintenance, so redir to homepage.
 */
export default function ClientMaintenanceCheck() {
  const {
    checkFeatureFlag, // An instance of FeatureFlagsManager
  } = useFeatureFlags();

  useEffect(() => {
    if (
      !checkFeatureFlag("searchOff") &&
      !checkFeatureFlag("opportunityOff") &&
      checkFeatureFlag("authOn")
    ) {
      // This piece does not seem reliable, maybe due to re-render or rehyrdrate cylces where we redirect away before the actual client live feature flag is set
      // redirect("/", RedirectType.push);
    }
  }, [checkFeatureFlag]);

  return null;
}
