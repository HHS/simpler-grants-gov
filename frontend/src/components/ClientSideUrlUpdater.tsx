"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { useRouter } from "next/navigation";
import { useEffect } from "react";

// This should be used as a way to force a url change from a server component
// supports updating:
// - the full url
// - an individual query param
// - the query string
// Note that if you want to zero out the query or param value, you should explicitly pass "" (empty string)
export function ClientSideUrlUpdater({
  param,
  value,
  url,
  query,
}: {
  param?: string;
  value?: string;
  url?: string;
  query?: string;
}) {
  const router = useRouter();
  const { updateQueryParams, searchParams } = useSearchParamUpdater();
  // if query has been passed in, use that, otherwise pass through the current query
  const updatedQuery = query !== undefined ? query : searchParams.get("query");

  useEffect(() => {
    if (url) {
      router.push(url);
    }
  }, [url, router]);
  useEffect(() => {
    if (param && value !== undefined) {
      updateQueryParams(value, param, updatedQuery);
    }
  }, [param, value, updatedQuery, updateQueryParams]);
  return <></>;
}
