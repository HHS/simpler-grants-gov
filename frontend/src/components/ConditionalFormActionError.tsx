import { ErrorMessage } from "@trussworks/react-uswds";

export function ConditionalFormActionError<
  T extends { [key: string]: string[] },
>({ errors, fieldName }: { errors?: T; fieldName: keyof T }) {
  if (!errors) {
    return;
  }
  const error = errors[fieldName];
  if (error) {
    return <ErrorMessage className="maxw-mobile-lg">{error[0]}</ErrorMessage>;
  }
}
