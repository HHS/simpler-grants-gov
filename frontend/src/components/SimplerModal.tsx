import { useIsSSR } from "src/hooks/useIsSSR";

import { KeyboardEventHandler, ReactNode, RefObject } from "react";
import { Modal, ModalHeading, ModalRef } from "@trussworks/react-uswds";

import "react-dom";

/*

  SimplerModal

  Wrapper for the Truss Modal component that provides common functionality shared
  by all modals within the Simpler application.

  * avoids pre-render errors using isSSR hook
  * allows custom onClose functionality, which is not supported by Truss out of the box

*/

export function SimplerModal({
  modalRef,
  className,
  modalId,
  titleText,
  children,
  onKeyDown,
  onClose,
}: {
  modalRef: RefObject<ModalRef | null>;
  titleText?: string;
  modalId: string;
  className?: string;
  children: ReactNode;
  onKeyDown?: KeyboardEventHandler<HTMLDivElement>;
  onClose?: () => void;
}) {
  // The Modal component throws an error during SSR unless we specify that it should not "render to portal"
  // this hook allows us to opt out of that rendering behavior on the server
  const isSSR = useIsSSR();

  return (
    <Modal
      ref={modalRef}
      forceAction={false}
      className={className}
      aria-labelledby={`${modalId}-heading`}
      aria-describedby={`${modalId}-description`} // be sure that children includes a div with this id
      id={modalId}
      renderToPortal={!isSSR}
      onClick={(clickEvent) => {
        if (!onClose) {
          return;
        }
        const clickTarget = clickEvent.target as HTMLElement;
        if (clickTarget.className.includes("usa-modal__close")) {
          onClose();
        }
      }}
      onKeyDown={(keyEvent) => {
        if (onClose && keyEvent.key === "Escape") {
          onClose();
        }
        onKeyDown && onKeyDown(keyEvent);
      }}
    >
      {titleText && (
        <ModalHeading id={`${modalId}-heading`}>{titleText}</ModalHeading>
      )}
      {children}
    </Modal>
  );
}
