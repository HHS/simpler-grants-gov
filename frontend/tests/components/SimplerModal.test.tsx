import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { noop } from "lodash";

import { createRef } from "react";

import { SimplerModal } from "src/components/SimplerModal";

const mockUseIsSSR = jest.fn();
const createPortalSpy = jest.fn();

jest.mock("src/hooks/useIsSSR", () => ({
  useIsSSR: () => mockUseIsSSR() as unknown,
}));

jest.mock("react-dom", () => ({
  ...jest.requireActual<typeof import("react-dom")>("react-dom"),
  createPortal: (modal: unknown) => createPortalSpy(modal) as unknown,
}));

describe("SimplerModal", () => {
  beforeEach(() => {
    mockUseIsSSR.mockReturnValue(false);
    createPortalSpy.mockImplementation((modal) => modal as unknown);
  });
  afterEach(() => {
    jest.resetAllMocks();
  });
  it("matches snapshot", () => {
    const { container } = render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={noop}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );

    expect(container).toMatchSnapshot();
  });
  it("displays header", () => {
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={noop}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );

    expect(screen.getByText("title text")).toBeInTheDocument();
  });
  it("displays children", () => {
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={noop}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );

    expect(screen.getByText("content")).toBeInTheDocument();
    expect(screen.getByText("content")).toBeVisible();
    expect(createPortalSpy).toHaveBeenCalled();
  });
  it("does not render to portal if ssr", () => {
    mockUseIsSSR.mockReturnValue(true);
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={noop}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );

    expect(createPortalSpy).not.toHaveBeenCalled();
  });
  it("runs onClose handler on escape key", async () => {
    const onCloseMock = jest.fn();
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={onCloseMock}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );
    const modal = screen.getByRole("button");
    modal.focus();
    await userEvent.keyboard("{Escape}");

    expect(onCloseMock).toHaveBeenCalled();
  });
  it("runs onClose handler on 'x' button click", async () => {
    const onCloseMock = jest.fn();
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={noop}
        onClose={onCloseMock}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );
    const xButton = screen.getByLabelText("Close this window");
    await userEvent.click(xButton);

    expect(onCloseMock).toHaveBeenCalled();
  });
  it("runs onKeydown function on key down", async () => {
    const user = userEvent.setup();
    const keyHandlerMock = jest.fn();
    render(
      <SimplerModal
        modalRef={createRef()}
        titleText="title text"
        modalId="modal-id"
        className="modal-class"
        onKeyDown={keyHandlerMock}
        onClose={noop}
      >
        <div id="modal-id-description">content</div>
      </SimplerModal>,
    );
    const modal = screen.getByRole("button");
    modal.focus();
    await user.keyboard("!");
    expect(keyHandlerMock).toHaveBeenCalled();
  });
});
