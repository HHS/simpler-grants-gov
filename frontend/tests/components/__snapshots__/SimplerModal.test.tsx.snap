// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimplerModal matches snapshot 1`] = `
<div>
  <div
    aria-describedby="modal-id-description"
    aria-labelledby="modal-id-heading"
    class="usa-modal-wrapper is-hidden"
    data-force-action="false"
    id="modal-id"
    role="dialog"
  >
    <div
      aria-controls="modal-id"
      class="usa-modal-overlay"
      data-testid="modalOverlay"
    >
      <div
        class="usa-modal modal-class"
        data-force-action="false"
        data-testid="modalWindow"
        tabindex="-1"
      >
        <div
          class="usa-modal__content"
        >
          <div
            class="usa-modal__main"
          >
            <h2
              class="usa-modal__heading"
              id="modal-id-heading"
            >
              title text
            </h2>
            <div
              id="modal-id-description"
            >
              content
            </div>
          </div>
          <button
            aria-controls="modal-id"
            aria-label="Close this window"
            class="usa-button usa-modal__close"
            data-close-modal="true"
            data-testid="button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="usa-icon"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
